### ISSUE TITLE
Swashbuckle and virtual directories

### ISSUE THREAD

**Issue by Icydillic** [domaindrivendev/Swashbuckle.AspNetCore#856]
Hello,

I encounter exactly the same problem as it is already documented here
https://github.com/domaindrivendev/Swashbuckle/issues/305#issuecomment-333081577

I am using version 3.0 with the following configuration

```csharp
...
services.AddSwaggerGen(swaggerGenOptions =>
{
    swaggerGenOptions.SwaggerDoc("v1", new Info { Title = "v1 API", Version = "0.0.1" });
});
...
app.UseSwagger();
app.UseSwaggerUI(options =>
{
    options.SwaggerEndpoint("/virtual/swagger/v1/swagger.json", "v1 API 0.0.1");
});
...
```

When I call the url http://localhost/virtual/swagger/index.html
I can see the swagger ui and click links.

But the api calls are looking like this:
`curl -X GET "http://localhost/api/Application/settings" -H  "accept: application/json"`

As you can see the api calls are wrong here. They are missing the "/virtual" path.
I can't seem to figure out what I need to do to tell swagger to use a virtual directory. I already tried "RoutePrefix" and "RouteTemplate" but that all seems to only address the swagger.json file location and not the swagger.json path content.

Has anyone else ever tried to get this to work with virtual directories?

**Comment by domaindrivendev** [comment]
For this to work - you need `basePath` to be set to "/virtual" in the Swagger JSON that's returned by Swashbuckle. As per the following snippet, this should be picked up automatically:

https://github.com/domaindrivendev/Swashbuckle.AspNetCore/blob/v3.0.0/src/Swashbuckle.AspNetCore.Swagger/Application/SwaggerMiddleware.cs#L42

But, it doesn't seem to be working in your case - I suspect `httpContext.Request.PathBase` is empty but I'm not sure why. As a workaround, you can manually assign the `basePath` as follows:

```csharp
 app.UseSwagger(c =>
 {
     c.PreSerializeFilters.Add((swagger, request) => swagger.BasePath = "/virtual");
 });
```

**Comment by Pietie99** [comment]
If it helps anyone, this resolved virtual paths for me, and running localhost works as well

`app.UseSwagger(c =>
            {
                c.RouteTemplate = "api/{documentName}/swagger.json";
                c.PreSerializeFilters.Add((swagger, httpReq) =>
                {
                     swagger.Servers = new List<OpenApiServer> { new OpenApiServer { Url = $"{httpReq.Scheme}://{httpReq.Host.Value}{httpReq.PathBase.Value}" } };
                });
            });`

**Comment by t-ricci-digit** [comment]
![image](https://github.com/domaindrivendev/Swashbuckle.AspNetCore/assets/118282270/3c1378f5-ba86-4635-891a-fcb905a31faf)

PreSerializeFilters is not available for me using Swashbucle 6.5.0 on .NET 7.0
